engineapi/proxy.go

```diff
@@ -100,69 +323,78 @@ func (h *EngineClientProxyHandler) GetClient(im *longhorn.InstanceManager) (c En
 		return nil, errors.Errorf("BUG: failed to get proxy client due to missing instance manager")
 	}
 
-	log := getLoggerForProxy(h.logger, im.Spec.Image)
+	_, exist := h.GetHandledInstanceManagerImage(im.Spec.Image)
+	if !exist {
+		for id := 0; id < DefaultProxyClientPerImage; id++ {
+			h.NewHandleClient(im, id)
+		}
+	}
 
-	h.lock.Lock()
-	client, exist := h.Clients[im.Spec.Image]
-	h.lock.Unlock()
-	if exist {
-		err := client.Ping()
-		if err == nil {
-			return client, nil
+	var candidate EngineClientProxy
+	failCount := 0
+	for {
+		if failCount > ProxyClientMaxRetry {
+			return nil, errors.Errorf("exceed max retry to get %v proxy client", im.Name)
 		}
 
-		log.WithError(err).Debug("Resetting proxy connection")
-		client.Stop(im.Spec.Image)
+		lessUsed := 0
+		handledIMImage, _ := h.GetHandledInstanceManagerImage(im.Spec.Image)
+		for _, client := range handledIMImage.Clients {
+			candidate = client
+			lessUsed = candidate.GetInUse()
+			break
+		}
 
-		h.lock.Lock()
-		delete(h.Clients, im.Spec.Image)
-		h.lock.Unlock()
-	}
+		for _, client := range handledIMImage.Clients {
+			if err := checkClient(client); err != nil {
+				continue
+			}
 
-	return h.NewClient(log, im)
-}
+			inUse := client.GetInUse()
+			if inUse >= lessUsed {
+				continue
+			}
 
-func (h *EngineClientProxyHandler) NewClient(log logrus.FieldLogger, im *longhorn.InstanceManager) (EngineClientProxy, error) {
-	h.lock.Lock()
-	defer h.lock.Unlock()
+			lessUsed = inUse
+			candidate = client
+		}
 
-	if existClient, exist := h.Clients[im.Spec.Image]; exist {
-		// TODO: we do not delete the map to avoid nil memory panic, so need to
-		// check the ping to find any concurrent NewClient. Fix this when
-		// refactoring to use goroutine for proxy client handling.
-		err := existClient.Ping()
-		if err == nil {
-			log.Debugf("Proxy client already created")
-			return existClient, nil
+		failCount++
+		if err := checkClient(candidate); err != nil {
+			clientLog := candidate.GetLogger()
+			clientLog.WithError(err).Debugf("Retry(%v) to get client", failCount)
+		} else {
+			candidate.SetInUse(true)
+			return candidate, nil
 		}
```

Comment:
Can we create another method to handle to find out a less used client?

Reply:
Addressed.